src/Entity/Form/ApiDocReimportSpecForm.php

```diff
@@ -0,0 +1,118 @@
+
+  /**
+   * The messenger service.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * The ApiDoc spec fetcher service.
+   *
+   * @var \Drupal\apigee_api_catalog\SpecFetcherInterface
+   */
+  protected $specFetcher;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected $operation = 'reimport_spec';
+
+  /**
+   * Constructs a ApiDocReimportSpecForm object.
+   *
+   * @param \Drupal\Core\Entity\EntityRepositoryInterface $entity_repository
+   *   The entity repository service.
+   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface $entity_type_bundle_info
+   *   The entity type bundle service.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   * @param \Drupal\apigee_api_catalog\SpecFetcherInterface $spec_fetcher
+   *   The ApiDoc spec fetcher service.
+   */
+  public function __construct(EntityRepositoryInterface $entity_repository, EntityTypeBundleInfoInterface $entity_type_bundle_info = NULL, TimeInterface $time = NULL, MessengerInterface $messenger, SpecFetcherInterface $spec_fetcher) {
+    parent::__construct($entity_repository, $entity_type_bundle_info, $time);
+    $this->messenger = $messenger;
+    $this->specFetcher = $spec_fetcher;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity.repository'),
+      $container->get('entity_type.bundle.info'),
+      $container->get('datetime.time'),
+      $container->get('messenger'),
+      $container->get('apigee_api_catalog.spec_fetcher')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQuestion() {
+    return $this->t('Are you sure you want to update the OpenAPI specification file from URL on API Doc %name?', [
+      '%name' => $this->entity->label(),
+    ]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCancelUrl() {
+    return new Url('entity.apidoc.collection');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDescription() {
+    return $this->t('This will replace the current OpenAPI specification file.
+     This action cannot be undone.');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    /* @var \Drupal\apigee_api_catalog\Entity\ApiDocInterface $entity */
+    $entity = $this->getEntity();
+
+    $needs_save = $this->specFetcher->fetchSpec($entity);
+    if ($needs_save) {
+      if ($entity->getEntityType()->isRevisionable()) {
+        $entity->setNewRevision();
+      }
+      $entity->save();
+      $this->messenger()->addStatus($this->t('API Doc %label: imported the OpenAPI specification file from URL.', [
+        '%label' => $this->entity->label(),
+      ]));
+    }
+    // TODO: $needs_save doesn't tell us if something went wrong.
```

Comment:
This is exactly why I suggested earlier to throw an exception when file fetching or saving fails instead of returning a boolean indicator value which is not the OOP solution.

Reply:
Yes, this can be addressed in a followup.

Reply:
This TODO was added so it can be changed in a follow up task.

Reply:
Which is the related follow-up task to this? Probably it would be better to create it as a GH issue, because that could be referenced in this PR.

Reply:
https://github.com/apigee/apigee-api-catalog-drupal/issues/14